operator/pkg/utils/pki/pki.go

```diff
@@ -0,0 +1,196 @@
+	return
+}
+
+func validCerts(certBytes, keyBytes []byte) (*x509.Certificate, crypto.Signer, error) {
+	cert, err := certutil.ParseCertsPEM(certBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+	key, err := keyutil.ParsePrivateKeyPEM(keyBytes)
+	if err != nil {
+		return nil, nil, err
+	}
+	return cert[0], key.(crypto.Signer), nil
+}
+
+func encode(cert *x509.Certificate, key crypto.Signer) (certBytes, keyBytes []byte) {
+	certBytes = pem.EncodeToMemory(&pem.Block{
+		Type:  "CERTIFICATE",
+		Bytes: cert.Raw,
+	})
+	keyBytes = pem.EncodeToMemory(&pem.Block{
+		Type:  "RSA PRIVATE KEY",
+		Bytes: x509.MarshalPKCS1PrivateKey(key.(*rsa.PrivateKey)),
+	})
+	return
+}
+
+func privateKeyAndCertificate(config *CertConfig, caCertBytes, caKeyBytes []byte) (*x509.Certificate, crypto.Signer, error) {
+	// verify the existing certs if valid use
+	cert, key, err := validCerts(config.ExistingCert, config.ExistingKey)
+	if err == nil {
+		zap.S().Infof("Reusing existing certs for %s", config.CommonName)
+		return cert, key, nil
+	}
+	// create private key, defaults to x509.RSA
+	if key, err = rsa.GenerateKey(cryptorand.Reader, rsaKeySize); err != nil {
+		return nil, nil, fmt.Errorf("unable to create private key while generating CA certificate, %w", err)
+	}
+	if cert, err = signedCert(config.Config, key, caCertBytes, caKeyBytes); err != nil {
+		return nil, nil, fmt.Errorf("creating signed cert, %w", err)
+	}
+	return cert, key, nil
+}
+
+// signedCert creates a signed certificate using the given CA certificate and key
+func signedCert(cfg *certutil.Config, key crypto.Signer, caCertBytes, caKeyBytes []byte) (*x509.Certificate, error) {
+	caCert, caKey, err := validCerts(caCertBytes, caKeyBytes)
+	if err != nil {
+		return nil, fmt.Errorf("root CA is invalid, %w", err)
+	}
+	serial, err := cryptorand.Int(cryptorand.Reader, new(big.Int).SetInt64(math.MaxInt64))
+	if err != nil {
+		return nil, err
+	}
+	if len(cfg.CommonName) == 0 {
+		return nil, errors.New("commonName is missing")
+	}
+	keyUsage := x509.KeyUsageKeyEncipherment | x509.KeyUsageDigitalSignature
+	removeDuplicateAltNames(&cfg.AltNames)
+	certTmpl := x509.Certificate{
+		Subject: pkix.Name{
+			CommonName:   cfg.CommonName,
+			Organization: cfg.Organization,
+		},
+		DNSNames:              cfg.AltNames.DNSNames,
+		IPAddresses:           cfg.AltNames.IPs,
+		SerialNumber:          serial,
+		NotBefore:             caCert.NotBefore,
+		NotAfter:              time.Now().Add(CertificateValidity).UTC(),
+		KeyUsage:              keyUsage,
+		ExtKeyUsage:           cfg.Usages,
+		BasicConstraintsValid: true,
+		IsCA:                  false,
+	}
+	certDERBytes, err := x509.CreateCertificate(cryptorand.Reader, &certTmpl, caCert, key.Public(), caKey)
+	if err != nil {
+		return nil, err
+	}
+	return x509.ParseCertificate(certDERBytes)
+}
+
+// removeDuplicateAltNames removes duplicate items in altNames.
+func removeDuplicateAltNames(altNames *certutil.AltNames) {
```

Comment:
Instead of a mutating function, consider an immutable approach

`func uniqueAltNames(altNames *certutil.AltNames) certutil.AltNames {}`